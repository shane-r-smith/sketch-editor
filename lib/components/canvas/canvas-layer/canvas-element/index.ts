export * from "./pen";
export * from "./eraser";

export * from "./canvas-element";
export * from "./canvas-element.types";
