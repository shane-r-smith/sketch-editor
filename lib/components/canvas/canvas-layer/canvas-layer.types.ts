import type { Layer } from "../../../domain";
import type { SketchElement } from "../../../domain/sketch-element";

export interface CanvasLayerProps {
  layer?: Layer;
  drawingElement?: SketchElement;
}
