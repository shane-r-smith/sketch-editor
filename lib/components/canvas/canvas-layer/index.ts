export * from "./canvas-element";
export * from "./canvas-layer";
export * from "./canvas-layer.types";
