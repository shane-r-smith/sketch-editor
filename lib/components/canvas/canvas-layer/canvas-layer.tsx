import type { SketchElement } from "../../../domain";
import { CanvasElement } from "./canvas-element";
import type { CanvasLayerProps } from "./canvas-layer.types";
import { Layer as KonvaLayer } from "react-konva";

export function CanvasLayer({ layer, drawingElement }: CanvasLayerProps) {
  if (!layer) {
    return null;
  }

  if (!layer.visible) {
    return null;
  }

  return (
    <KonvaLayer opacity={layer.opacity}>
      {layer.elements.map((element: SketchElement, i) => (
        <CanvasElement
          key={i}
          element={element}
          offsetY={layer.index}
          opacity={layer.opacity}
        />
      ))}
      {drawingElement && (
        <CanvasElement
          element={drawingElement}
          offsetY={layer.index}
          opacity={layer.opacity}
        />
      )}
    </KonvaLayer>
  );
}
