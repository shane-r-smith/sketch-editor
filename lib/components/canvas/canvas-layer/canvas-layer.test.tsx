import { render } from "@testing-library/react";
import { CanvasLayer } from "./canvas-layer";
import type { CanvasLayerProps } from "./canvas-layer.types";
import { vi } from "vitest";
import type { CanvasElementProps } from "./canvas-element";
import type { LayerConfig } from "konva/lib/Layer";
import type { Layer } from "../../../domain";

// Mock CanvasElement and KonvaLayer
vi.mock("./canvas-element", () => ({
  CanvasElement: (props: CanvasElementProps) => (
    <div data-testid="canvas-element">
      {JSON.stringify(props, undefined, 2)}
    </div>
  ),
}));

vi.mock("react-konva", () => ({
  Layer: ({ children, opacity }: LayerConfig) => (
    <div data-testid="konva-layer" data-opacity={opacity}>
      {children}
    </div>
  ),
}));

describe("<CanvasLayer />", () => {
  const layer: Layer = {
    id: "layer-1",
    name: "layer-1",
    opacity: 1,
    elements: [
      {
        tool: "PEN",
        points: [0, 0, 10, 10],
        stroke: "#000000",
        strokeWidth: 2,
        strokeColourType: "HEX",
      },
      {
        tool: "PEN",
        points: [0, 10, 10, 0],
        stroke: "#000000",
        strokeWidth: 2,
        strokeColourType: "HEX",
      },
      {
        tool: "ERASER",
        points: [0, 10, 10, 0],
        stroke: "#000000",
        strokeWidth: 2,
      },
    ],
    visible: true,
    locked: false,
  };

  const defaultProps: CanvasLayerProps = {
    layer: layer,
    index: 0,
  };

  it("renders correctly and matches snapshot", () => {
    const { container } = render(<CanvasLayer {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("renders null if layer.visible is false", () => {
    const { container } = render(
      <CanvasLayer
        layer={{ ...layer, visible: false }}
        index={0}
        drawingElement={undefined}
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the drawing element", () => {
    const { getAllByTestId } = render(
      <CanvasLayer
        {...defaultProps}
        drawingElement={{
          tool: "PEN",
          points: [5, 5, 15, 15],
          stroke: "#FF00FF",
          strokeWidth: 3,
          strokeColourType: "HEX",
        }}
      />
    );
    const elements = getAllByTestId("canvas-element");
    // 3 from layer.elements + 1 from prop 'element'
    expect(elements).toHaveLength(4);
    expect(elements[3]).toMatchSnapshot();
  });
});
