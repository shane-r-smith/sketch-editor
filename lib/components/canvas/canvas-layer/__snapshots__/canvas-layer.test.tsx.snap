// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<CanvasLayer /> > renders correctly and matches snapshot 1`] = `
<div>
  <div
    data-opacity="1"
    data-testid="konva-layer"
  >
    <div
      data-testid="canvas-element"
    >
      {
  "element": {
    "tool": "PEN",
    "points": [
      0,
      0,
      10,
      10
    ],
    "stroke": "#000000",
    "strokeWidth": 2,
    "strokeColourType": "HEX"
  },
  "offsetY": 0,
  "opacity": 1
}
    </div>
    <div
      data-testid="canvas-element"
    >
      {
  "element": {
    "tool": "PEN",
    "points": [
      0,
      10,
      10,
      0
    ],
    "stroke": "#000000",
    "strokeWidth": 2,
    "strokeColourType": "HEX"
  },
  "offsetY": 0,
  "opacity": 1
}
    </div>
    <div
      data-testid="canvas-element"
    >
      {
  "element": {
    "tool": "ERASER",
    "points": [
      0,
      10,
      10,
      0
    ],
    "stroke": "#000000",
    "strokeWidth": 2
  },
  "offsetY": 0,
  "opacity": 1
}
    </div>
  </div>
</div>
`;

exports[`<CanvasLayer /> > renders null if layer is undefined 1`] = `<div />`;

exports[`<CanvasLayer /> > renders null if layer.visible is false 1`] = `<div />`;

exports[`<CanvasLayer /> > renders the drawing element 1`] = `
<div
  data-testid="canvas-element"
>
  {
  "element": {
    "tool": "PEN",
    "points": [
      5,
      5,
      15,
      15
    ],
    "stroke": "#FF00FF",
    "strokeWidth": 3,
    "strokeColourType": "HEX"
  },
  "offsetY": 0,
  "opacity": 1
}
</div>
`;
